File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/account/AccountViewModel.kt
package me.nathanfallet.bdeensisa.features.account

import android.app.Application
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.extensions.generateQRCode
import me.nathanfallet.bdeensisa.models.Ticket
import me.nathanfallet.bdeensisa.models.User
import me.nathanfallet.bdeensisa.models.UserToken

class AccountViewModel(
    application: Application,
    code: String?,
    token: String?,
    id: String?,
    val saveToken: (UserToken) -> Unit,
    val showAccount: () -> Unit
) : AndroidViewModel(application) {

    // Properties

    private var qrCode = MutableLiveData<Bitmap>()
    private val tickets = MutableLiveData<List<Ticket>>()

    // Getters

    fun getQrCode(): LiveData<Bitmap> {
        return qrCode
    }

    fun getTickets(): LiveData<List<Ticket>> {
        return tickets
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "account")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "AccountView")
        }

        fetchData(token, id)
        code?.let {
            authenticate(it)
        }
    }

    fun fetchData(token: String?, id: String?) {
        if (token == null || id == null) {
            return
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getUserTickets(token, id).let {
                    tickets.value = it
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun launchLogin() {
        val browserIntent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse(
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication()))
                    .apiService().authenticationUrl
            )
        )
        browserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        ContextCompat.startActivity(getApplication(), browserIntent, null)
    }

    fun authenticate(code: String) {
        viewModelScope.launch {
            try {
                val token = SharedCacheService.getInstance(DatabaseDriverFactory(getApplication()))
                    .apiService().authenticate(code)
                saveToken(token)
                showAccount()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun generateQrCode(user: User) {
        viewModelScope.launch {
            val code = "bdeensisa://users/${user.id}".generateQRCode()
            qrCode.value = code
        }
    }

    fun launchPayment(
        token: String?,
        shopItemType: String,
        shopItemId: String,
        itemId: String
    ) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                val response = SharedCacheService.getInstance(DatabaseDriverFactory(getApplication()))
                    .apiService().getShopItemPayment(token, shopItemType, shopItemId, itemId)
                val browserIntent = Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse(response.url)
                )
                browserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                ContextCompat.startActivity(getApplication(), browserIntent, null)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

}

File: ios/SuiteBDE/Features/Account/AccountViewModel.swift
//
//  AccountViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 31/01/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import WidgetKit
import shared

class AccountViewModel: ObservableObject {
    
    let saveToken: (UserToken) -> Void
    
    @Published var url: String?
    @Published var error: String?
    @Published var qrcode: UIImage?
    @Published var tickets = [Ticket]()
    
    init(saveToken: @escaping (UserToken) -> Void) {
        self.saveToken = saveToken
    }
    
    func onAppear(token: String?, id: String?) {
        AnalyticsService.shared.log(.screenView(screenName: "account", screenClass: "AccountView"))
        
        fetchData(token: token, id: id)
    }
    
    func fetchData(token: String?, id: String?) {
        guard let token, let id else {
            return
        }
        Task {
            let tickets = try await CacheService.shared.apiService().getUserTickets(token: token, id: id)
            DispatchQueue.main.async {
                self.tickets = tickets
            }
        }
    }
    
    func launchLogin() {
        url = CacheService.shared.apiService().authenticationUrl
    }
    
    func onOpenUrl(url: URL) {
        guard url.host == "authorize" else { return }
        self.url = nil
        Task {
            let userToken = try await CacheService.shared.apiService().authenticate(code: url.absoluteString)
            DispatchQueue.main.async {
                self.saveToken(userToken)
                WidgetCenter.shared.reloadAllTimelines()
            }
            guard let fcmToken = StorageService.userDefaults?.value(forKey: "fcmToken") as? String else {
                return
            }
            try await CacheService.shared.apiService().sendNotificationToken(
                token: userToken.token,
                notificationToken: fcmToken
            )
        }
    }
    
    func generateQRCode(user: User) {
        Task {
            let code = "bdeensisa://users/\(user.id)".generateQRCode()
            DispatchQueue.main.async {
                self.qrcode = code
            }
        }
    }
    
    func launchPayment(
        token: String?,
        shopItemType: String,
        shopItemId: String,
        itemId: String
    ) {
        guard let token else {
            return
        }
        Task {
            let response = try await CacheService.shared.apiService()
                .getShopItemPayment(token: token, type: shopItemType, id: shopItemId, itemId: itemId)
            DispatchQueue.main.async {
                self.url = response.url
            }
        }
    }
    
}
